oak_restricted_kernel_api/src/dice.rs

```diff
@@ -0,0 +1,71 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::syscall::read;
+use oak_crypto::{
+    encryptor::EncryptionKeyProvider,
+    hpke::{Deserializable, PrivateKey, PublicKey},
+};
+use oak_dice::evidence::{
+    RestrictedKernelDiceData, P256_PRIVATE_KEY_SIZE, X25519_PRIVATE_KEY_SIZE,
+};
+use oak_remote_attestation::proto::oak::attestation::v1::Evidence;
+use oak_restricted_kernel_interface::DICE_DATA_FD;
+use p256::ecdsa::SigningKey;
+use zerocopy::{AsBytes, FromZeroes};
+
+/// Wrapper for DICE evidence and application private keys.
+pub struct DiceWrapper {
+    pub evidence: Evidence,
+    pub encryption_key: EncryptionKeyProvider,
+    pub signing_key: SigningKey,
+}
+
+/// Get the DICE evidence and application private keys from the Restricted Kernel
+pub fn get_dice_evidence_and_keys() -> anyhow::Result<DiceWrapper> {
+    let dice_data = get_restricted_kernel_dice_data()?;
+    let evidence: Evidence = dice_data.evidence.try_into()?;
+    let private_key = PrivateKey::from_bytes(
+        &dice_data.application_private_keys.encryption_private_key[..X25519_PRIVATE_KEY_SIZE],
+    )
+    .map_err(|error| anyhow::anyhow!("couldn't deserialize private key: {}", error))?;
+    let application_keys = evidence
+        .application_keys
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no application keys"))?;
+    let public_key = PublicKey::from_bytes(application_keys.encryption_public_key_certificate())
```

Comment:
I think we could derive the public key from the private key as part of `new_from_keys` function, so that we won't depend on the evidence, and potentially could split the API in the future

Reply:
I don't think we should be deriving keys. The public key must be bound to the evidence for clients to trust it, so we should be consistent in using the public key from the evidence on the server.